### PR TITLE
Propose vulners.nse to official nmap repo.

### PR DESCRIPTION
**Pull request by GMedian**
The proposed script is the latest stable release of vulnersCom/nmap-vulners github repo.

### CODE REVIEW

**Comment by cldrn** [comment]
What's the query limit per IP?

I've looked into this before but at the time we had a working version of
vulscan which does the same thing with a local database so we didnt have to
deal with any privacy issues too.

However, The script works as expected so besides the expected additional
server load on the service which you obviously are OK with, it might be
worth adding it as an external option or even merging into vulscan.nse

**Comment by vulnersCom** [comment]
At the moment it's about 1000rps. No one user even gets close))
Has not figured performance issues yet. Caching and direct queries works fine.

Unfortunately no way to use local db, it's about 250gb in elastics and API works querying it on the fly.

What do you mean by "external option"?

File: scripts/vulners.nse [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+-- |     CVE-2002-0029        7.5        https://vulners.com/cve/CVE-2002-0029
+-- |     CVE-2015-5986        7.1        https://vulners.com/cve/CVE-2015-5986
+-- |     CVE-2010-3615        5.0        https://vulners.com/cve/CVE-2010-3615
+-- |     CVE-2006-0987        5.0        https://vulners.com/cve/CVE-2006-0987
+-- |     CVE-2014-3214        5.0        https://vulners.com/cve/CVE-2014-3214
+--
+
+author = 'gmedian AT vulners DOT com'
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"vuln", "safe"}
```
Comment (p-l-) [inline scripts/vulners.nse]: If I understand well, this script **must** also be categorized as "external".

File: scripts/vulners.nse [hunk 2]
```diff
@@ -0,0 +1,214 @@
+description = [[
+For each available cpe it prints the known vulns (links to the correspondent info).
+
+Its work is pretty simple:
+- work only when some software version is identified for an open port
+- take all the known cpe for that software (from the standard nmap output)
+- ask whether some known vulns exist for that cpe
```
Comment (p-l-) [inline scripts/vulners.nse]: You probably want to mention here that it connects to a remote service.

File: scripts/vulners.nse (hunk 1 above)
Comment (vulnersCom) [reply scripts/vulners.nse]: Got it.

File: scripts/vulners.nse (hunk 2 above)
Comment (vulnersCom) [reply scripts/vulners.nse]: Definitely yes. Will refactor the description.

**Comment by GMedian** [comment]
@vulnersCom @cldrn @p-l- 
Thank you for your time, I have updated the script to explicitly mention the remote API.

**Comment by vulnersCom** [comment]
Is it ok now?

**Comment by b4ldr** [comment]
message for the devs, the https://vulners.com/products web site explicitly mentions nmap scrips and the API seems to be free and open (https://vulners.com/docs) and all links from the main site point to the @vulnersCom github account so i think we think permission to use the API has been granted. anything else holding this up, it looks like a really nice addition?

**Comment by vulnersCom** [comment]
Hi fellows.
Anything we can do more?

**Comment by aidylewis** [comment]
I use this script as a useful addition to the vuln category. I'd be nice for it to be in the main scripts.

**Comment by vulnersCom** [comment]
Hi devs! Any movements here?

**Comment by aidylewis** [comment]
Bump.

**Comment by aidylewis** [comment]
We use this script quite extensively in a large media company and it would be really nice if it was in the official scripts directory.

**Comment by aidylewis** [comment]
We use this script extensively and have built pretty printing, persistence and visualisation (which we will eventually open source). Can we please get this merged into the the nmap official scripts or is there something holding this back? Some users have a problem finding the **scripts** folder.

**Comment by vulnersCom** [comment]
Bump

**Comment by aidylewis** [comment]
Not sure if nmap needs a separate NSE chief maintainer and other lower down the chain. Some of those scripts are old, and this is where I see the innovation emerging. WDYT?  @vulnersCom

**Comment by dmiller-nmap** [comment]
@vulnersCom I'm working on getting this done with a few adjustments, but I can't seem to get any results. Every query is coming back with the following response:

```
HTTP/1.1 200 OK
Server: QRATOR
Date: Wed, 26 Jun 2019 03:46:45 GMT
Content-Type: application/json
Content-Length: 124
Connection: close
Vary: Accept-Encoding
X-Vulners-Ratelimit-Reqlimit: 300
X-Vulners-Ratelimit-Burstlimit: 50
X-Vulners-Ratelimit-Rate: 228.9639381797367
X-Vulners-View-Calltime: 60ms
X-Vulners-Calltime: 60ms
X-Vulners-Full-Calltime: 78ms
Set-Cookie: vulnersSession=6YYOKPTQAZNYK8MTROX1TITSZWILACFG0GC4LQOGQGLZ761ZI1ND62N72BQSVFBUT7GUE2MILYF1DDHKO784NJ3521BVHSKN7FWPSRCNBDWQ1ZCEUG6R0W2YT1VRZ00G:4q-_Xe2Kk2HzNAUT_dh88p73z9Q; Domain=.vulners.com; expires=Wed, 26 Jun 2019 03:46:55 GMT; HttpOnly; Max-Age=10; Path=/; Secure
Strict-Transport-Security: max-age=63072000; includeSubdomains; preload
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block

{   
  "result": "warning",
  "data": {
    "warning": "Nothing found for Burpsuite search request",
    "errorCode": 401
  } 
}   
```

Is this expected? Is there a particular query (CPE, software name and version, etc.) that I could use to get a valid result for testing?

**Comment by vulnersCom** [comment]
@dmiller-nmap 
Here we go. Test examples:

https://vulners.com/api/v3/burp/software/?type=cpe&software=cpe:/a:mongodb:mongodb&version=2.2
https://vulners.com/api/v3/burp/software/?type=cpe&software=cpe:/a:cybozu:garoon&version=4.2.1

There is always 200 reply if server side is OK. But differs in JSON "result" var.

**Comment by dmiller-nmap** [comment]
Committed in 2056ef9 with further changes:

* ba530533db23c6186650f814b351cfc2bf3993c9 - Whitespace and NSEdoc formatting
* 86d448e31f9ce91b1fee857e172b32526e6a857b - Avoid excessive string concatenation
* 70f1cbba3d01bd2355018d97e0568ebcb9aac43b - Structured output
* a08be753fb67753992f651b5cdca98abc2274af8 - Enable vulners.nse to work during IPv6 scans
* fbcaa39fa2d27b1588619ad2b01713b2ae1559b3 - Document mincvss arg, ensure consistent ordering

**Pull request closed without merging**